File: examples/java/kora-java-crud-submodule/kora-java-crud-submodule-app/src/main/java/ru/tinkoff/kora/example/submodule/app/controller/VetDelegate.java
package ru.tinkoff.kora.example.submodule.app.controller;

import static ru.tinkoff.kora.example.submodule.openapi.http.server.api.VetApiResponses.*;

import ru.tinkoff.kora.common.Component;
import ru.tinkoff.kora.example.submodule.app.model.VetMapper;
import ru.tinkoff.kora.example.submodule.openapi.http.server.api.VetApiDelegate;
import ru.tinkoff.kora.example.submodule.openapi.http.server.model.MessageTO;
import ru.tinkoff.kora.example.submodule.openapi.http.server.model.VetCreateTO;
import ru.tinkoff.kora.example.submodule.openapi.http.server.model.VetUpdateTO;
import ru.tinkoff.kora.example.submodule.vet.service.VetService;

@Component
public final class VetDelegate implements VetApiDelegate {

    private final VetMapper vetMapper;
    private final VetService vetService;

    public VetDelegate(VetMapper vetMapper, VetService vetService) {
        this.vetMapper = vetMapper;
        this.vetService = vetService;
    }

    @Override
    public ListVetsApiResponse listVets() {
        var vets = vetService.findAll();
        var vetTOs = vets.stream()
                .map(vetMapper::asDTO)
                .toList();

        return new ListVetsApiResponse.ListVets200ApiResponse(vetTOs);
    }

    @Override
    public GetVetByIdApiResponse getVetById(long vetId) {
        if (vetId < 0) {
            return new GetVetByIdApiResponse.GetVetById400ApiResponse(malformedId(vetId));
        }

        var vet = vetService.findByID(vetId);
        if (vet.isPresent()) {
            var body = vetMapper.asDTO(vet.get());
            return new GetVetByIdApiResponse.GetVetById200ApiResponse(body);
        } else {
            return new GetVetByIdApiResponse.GetVetById404ApiResponse(notFound(vetId));
        }
    }

    @Override
    public AddVetApiResponse addVet(VetCreateTO vetCreateTO) {
        var vet = vetService.add(vetCreateTO.firstName(), vetCreateTO.lastName());
        var body = vetMapper.asDTO(vet);
        return new AddVetApiResponse.AddVet200ApiResponse(body);
    }

    @Override
    public UpdateVetApiResponse updateVet(long vetId, VetUpdateTO vetUpdateTO) {
        if (vetId < 0) {
            return new UpdateVetApiResponse.UpdateVet400ApiResponse(malformedId(vetId));
        }

        var updated = vetService.update(vetId, vetUpdateTO.firstName(), vetUpdateTO.lastName());
        if (updated.isPresent()) {
            var body = vetMapper.asDTO(updated.get());
            return new UpdateVetApiResponse.UpdateVet200ApiResponse(body);
        } else {
            return new UpdateVetApiResponse.UpdateVet404ApiResponse(notFound(vetId));
        }
    }

    @Override
    public DeleteVetApiResponse deleteVet(long vetId) {
        if (vetId < 0) {
            return new DeleteVetApiResponse.DeleteVet400ApiResponse(malformedId(vetId));
        }

        if (vetService.delete(vetId)) {
            return new DeleteVetApiResponse.DeleteVet200ApiResponse(new MessageTO("Successfully deleted Vet with ID: " + vetId));
        } else {
            return new DeleteVetApiResponse.DeleteVet404ApiResponse(notFound(vetId));
        }
    }

    private static MessageTO notFound(long vetId) {
        return new MessageTO("Vet not found for ID: " + vetId);
    }

    private static MessageTO malformedId(long vetId) {
        return new MessageTO("Vet malformed ID: " + vetId);
    }
}

File: examples/kotlin/kora-kotlin-crud-submodule/kora-kotlin-crud-submodule-app/src/main/kotlin/ru/tinkoff/kora/kotlin/example/submodule/app/controller/VetDelegate.kt
package ru.tinkoff.kora.kotlin.example.submodule.app.controller

import ru.tinkoff.kora.common.Component
import ru.tinkoff.kora.kotlin.example.submodule.app.model.VetMapper
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.api.VetApiDelegate
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.api.VetApiResponses
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.model.MessageTO
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.model.VetCreateTO
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.model.VetUpdateTO
import ru.tinkoff.kora.kotlin.example.submodule.vet.service.VetService

@Component
class VetDelegate(
    private val vetMapper: VetMapper,
    private val vetService: VetService,
) : VetApiDelegate {
    override fun listVets(): VetApiResponses.ListVetsApiResponse {
        val vets = vetService.findAll().map(vetMapper::asDTO)
        return VetApiResponses.ListVetsApiResponse.ListVets200ApiResponse(vets)
    }

    override fun getVetById(vetId: Long): VetApiResponses.GetVetByIdApiResponse {
        if (vetId < 0) {
            return VetApiResponses.GetVetByIdApiResponse.GetVetById400ApiResponse(malformedId(vetId))
        }

        val vet = vetService.findByID(vetId)
        return if (vet != null) {
            VetApiResponses.GetVetByIdApiResponse.GetVetById200ApiResponse(vetMapper.asDTO(vet))
        } else {
            VetApiResponses.GetVetByIdApiResponse.GetVetById404ApiResponse(notFound(vetId))
        }
    }

    override fun addVet(vetCreateTO: VetCreateTO): VetApiResponses.AddVetApiResponse {
        val vet = vetService.add(vetCreateTO.firstName, vetCreateTO.lastName)
        return VetApiResponses.AddVetApiResponse.AddVet200ApiResponse(vetMapper.asDTO(vet))
    }

    override fun updateVet(vetId: Long, vetUpdateTO: VetUpdateTO): VetApiResponses.UpdateVetApiResponse {
        if (vetId < 0) {
            return VetApiResponses.UpdateVetApiResponse.UpdateVet400ApiResponse(malformedId(vetId))
        }

        val updated = vetService.update(vetId, vetUpdateTO.firstName, vetUpdateTO.lastName)
        return if (updated != null) {
            VetApiResponses.UpdateVetApiResponse.UpdateVet200ApiResponse(vetMapper.asDTO(updated))
        } else {
            VetApiResponses.UpdateVetApiResponse.UpdateVet404ApiResponse(notFound(vetId))
        }
    }

    override fun deleteVet(vetId: Long): VetApiResponses.DeleteVetApiResponse {
        if (vetId < 0) {
            return VetApiResponses.DeleteVetApiResponse.DeleteVet400ApiResponse(malformedId(vetId))
        }

        return if (vetService.delete(vetId)) {
            VetApiResponses.DeleteVetApiResponse.DeleteVet200ApiResponse(MessageTO("Successfully deleted Vet with ID: $vetId"))
        } else {
            VetApiResponses.DeleteVetApiResponse.DeleteVet404ApiResponse(notFound(vetId))
        }
    }

    private fun notFound(vetId: Long): MessageTO = MessageTO("Vet not found for ID: $vetId")
    private fun malformedId(vetId: Long): MessageTO = MessageTO("Vet malformed ID: $vetId")
}
